File: client/src/simulation/Entity.ts
import {EntityComponentData, EntityData} from "./EntityData";
import {Simulation} from "./Simulation";
import {EntityComponent} from "./EntityComponent";

export class Entity {
  data: EntityData

  readonly components: EntityComponent<any>[]
  readonly componentByServerType: Record<string, EntityComponent<any>>

  readonly simulation: Simulation;
  readonly entityId: string

  constructor(
    simulation: Simulation,
    initialData: EntityData
  ) {
    this.data = initialData
    this.simulation = simulation
    this.components = initialData.components.map(componentData => new EntityComponent<any>(this, componentData))
    this.entityId = initialData.entityId

    const componentByServerType: Record<string, EntityComponent<any>> = {}
    for (let component of this.components) {
      componentByServerType[component.data.type] = component
    }

    this.componentByServerType = componentByServerType
  }

  getComponentDataOrNull<T extends EntityComponentData>(serverTypeName: string): T | null {
    const component = this.getComponentOrNull<T>(serverTypeName)

    if (component == null) {
      return null
    }

    return component.data
  }

  getComponentData<T extends EntityComponentData>(serverTypeName: string): T {
    return this.getComponent<T>(serverTypeName).data
  }

  getComponentOrNull<T extends EntityComponentData>(serverTypeName: string): EntityComponent<T> | null {
    return this.componentByServerType[serverTypeName]
  }

  getComponent<T extends EntityComponentData>(serverTypeName: string): EntityComponent<T> {
    const result = this.getComponentOrNull<T>(serverTypeName)

    if (result == null) {
      throw new Error(`Component not found on entity: ${serverTypeName} (${this.entityId})`)
    }

    return result
  }

  handleNewComponentData(data: EntityData) {
    this.data = data
  }
}

File: simulation-server/src/main/kotlin/botfarm/simulationserver/simulation/Entity.kt
package botfarm.simulationserver.simulation

import kotlin.reflect.KClass
import kotlin.reflect.full.isSubclassOf



class Entity(
   data: EntityData,
   val simulation: Simulation
) {
   val entityId = data.entityId

   var destroyedAtTime: Double? = null
      private set


   private val mutableComponents: MutableList<EntityComponent<*>> = data.components.map {
      EntityComponent<EntityComponentData>(
         data = it,
         entity = this,
         componentDataClass = it.javaClass.kotlin
      )
   }.toMutableList()

   val components: List<EntityComponent<*>> = this.mutableComponents

   fun start(
      startContext: StartContext,
      simulation: Simulation
   ) {
      this.mutableComponents.forEach {
         it.start(
            startContext = startContext,
            simulation = simulation,
            entity = this
         )
      }
   }



   fun <T : EntityComponentData> getComponentOrNull(componentType: KClass<T>): EntityComponent<T>? {
      val untypedResult = this.mutableComponents.find {
         it.data::class.isSubclassOf(componentType)
      }

      if (untypedResult != null) {
         @Suppress("UNCHECKED_CAST")
         return untypedResult as EntityComponent<T>
      } else {
         return null
      }
   }

   inline fun <reified T : EntityComponentData> getComponentOrNull(): EntityComponent<T>? {
      return this.getComponentOrNull(T::class)
   }

   fun <T : EntityComponentData> getComponent(componentType: KClass<T>): EntityComponent<T> {
      return this.getComponentOrNull(componentType)
         ?: throw Exception("Entity does not have component for type: $componentType")
   }

   inline fun <reified T : EntityComponentData> getComponent(): EntityComponent<T> {
      return this.getComponent<T>(T::class)
   }

   fun destroy() {
      this.destroyedAtTime = this.simulation.getCurrentSimulationTime()
      this.simulation.handleEntityDestroyed(entity = this)
   }

   fun buildData(): EntityData {
      return EntityData(
         entityId = this.entityId,
         components = this.mutableComponents.map {
            it.data
         }
      )
   }

   fun stop() {
      this.mutableComponents.forEach {
         it.stop()
      }
   }
}